README.md
# Livefield.js

http://github.com/withassociates/livefield.js

A simple live-lookup ui library.

## Usage

As a basic use case, let's say we have an input in a cms that lets users
specify which page a thumbnail ought to link to.

We have an action on the server `/pages/search` that will find matching
pages and return them as JSON.

This is how we'd implement livefield:

    <!-- input to bind to -->
    <input
      type="text"
      name="link_path"
      id="my-input"
      data-store="/pages/search/{{query}}"
      data-template="#link-template"
    />

    <!-- template for result options -->
    <script type="template/handlebars" id="link-template">
      <li data-value="{{path}}">
        <span class="name">{{name}}</span>
        <span class="path">{{path}}</span>
      </li>
    </script>

    <!-- and activate livefield -->
    <script>
      $('#my-input').livefield();
    </script>

### Data-store

Your input needs `data-store` to specify the lookup URL. This can be
plain text or a handlebars template. If it's a template, it will
receive just the `query` paramters.

### Data-template

Your input also needs `data-template`. This should be a jQuery-compatible
selector for the element containing the template. Typically, this will
be in a `script` tag as in the example above.

### Data-value

Your result option template needs to have the `data-value` option.
The value of this attribute will be used to populate your input.

## Dependencies

* [jQuery](http://jquery.com/) ~> 1.5
* [Handlebars](http://handlebars.strobeapp.com/) ~> 1.0

## Contributors

* jamie@withassociates.com

## License

(The MIT License)

Copyright (c) 2011 With Associates

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

CHANGELOG.md
# 0.11.0 / 2012-02-03

* Features

  * Triggers "change" event at appropriate moments

# 0.10.0 / 2011-06-29

* Features

  * Results drop-up if there's insufficient space below the input

# 0.9.0 / 2011-06-14

* Features

  * Beta release


spec/spec_helper.js
function createContainer() {
  if ($('#spec').length === 0) {
    $('<div id="spec" />').appendTo('body');
  }
}

function clearContainer() {
  $('#spec').html('');
}

function clear() {
  clearContainer();
  $('.livefield-results').remove();
}

function createInput() {
  var $input =
    $('<input />').
    attr('type'         , 'text').
    attr('id'           , 'my-input').
    attr('data-store'   , 'spec/fixtures/my_store.json').
    attr('data-template', '#livefield-result-template');

  $('#spec').append($input);

  return $input;
}

function typesCharacter(char) {
  runs(function() {
    $('#my-input').focus();
    $('#my-input').val(char || 'a').trigger('keydown');
  });
  waits(100);
}

function clearsInput() {
  runs(function() {
    $('#my-input').val('').trigger('keydown');
  });
  waits();
}

function pressesKey(code) {
  runs(function() {
    var e = $.Event('keydown');

    e.which = {
      'up'    : 38,
      'down'  : 40,
      'esc'   : 27,
      'enter' : 13
    }[code];

    $('#my-input').trigger(e);
  });
  waits(100);
}

function blursInput() {
  runs(function() {
    this.$input.blur();
  });
  waits(100);
}

function focusesInput() {
  runs(function() {
    this.$input.focus();
  });
  waits(100);
}

LivefieldMatchers = {

  toHaveClass: function(klass) {
    return $(this.actual).hasClass(klass);
  },

  toBeHighlighted: function() {
    return $(this.actual).hasClass('livefield-highlighted');
  },

  toHaveFocus: function() {
    return $(this.actual)[0] === $('*:focus')[0];
  },

  toMatchElements: function(count) {
    if (count === undefined) {
      return $(this.actual).length > 0;
    } else {
      return $(this.actual).length === count;
    }
  }

}


spec/livefield/store_spec.js
describe('Livefield.Store', function() {

  describe('#find', function() {
    it('calls jQuery.getJSON', function() {
      var store = new Livefield.Store({ url: 'spec/fixtures/my_store.json' });
      spyOn(jQuery, 'getJSON').andCallThrough();
      store.find('foo', function() {});
      expect(jQuery.getJSON).toHaveBeenCalledWith('spec/fixtures/my_store.json');
    });

    it('calls callback with data', function() {
      var store = new Livefield.Store({ url: 'spec/fixtures/my_store.json' });
      var callback = jasmine.createSpy('callback');
      runs(function() { store.find('foo', callback) });
      waits(100);
      runs(function() { expect(callback).toHaveBeenCalled() });
    });

    describe('with a templated url', function() {
      it('inserts the query into the template', function() {
        var store = new Livefield.Store({ url: '/pages/search/:query' });
        spyOn(jQuery, 'getJSON').andCallThrough();
        store.find('foo', function() {});
        expect(jQuery.getJSON).toHaveBeenCalledWith('/pages/search/foo');
      });
    });
  });

  describe('#queue', function() {
    beforeEach(function() {
      this.store = new Livefield.Store({ url: 'spec/fixtures/my_store.json' });
    });

    it('adds new requests to the queue', function() {
      var args = ['foo', function() {}];
      spyOn(this.store.queue, 'push').andCallThrough();
      this.store.find(args[0], args[1]);
      expect(this.store.queue.push).toHaveBeenCalledWith(args);
    });

  });

});


spec/livefield/layout_spec.js
describe('Livefield results list', function() {

  beforeEach(function() {
    createContainer();
    this.addMatchers(LivefieldMatchers);
    this.$input = createInput();
    this.$input.css({
      position: 'absolute',
      width: '750px',
      top: 0,
      bottom: 'auto'
    });
    this.controller = new Livefield.Controller({ input: '#my-input' });
    typesCharacter();
  });

  afterEach(function() {
    clear();
  });

  it('is positioned absolutely', function() {
    runs(function() {
      expect($('.livefield-results').css('position')).toEqual('absolute');
    });
  });

  it('is positioned directly below the input', function() {
    var inputBottom = this.$input.offset().top + this.$input.outerHeight();
    var resultsTop = $('.livefield-results').offset().top - parseInt($('.livefield-results').css('margin-top'), 10);
    expect(resultsTop).toEqual(inputBottom);
  });

  it('is aligned horizontally with the input', function() {
    runs(function() {
      this.$input.css('margin-left', 20);
      $(window).trigger('resize');
    });
    waits();
    runs(function() {
      var inputLeft = this.$input.offset().left;
      var resultsLeft = $('.livefield-results').offset().left;
      expect(resultsLeft).toEqual(inputLeft);
    });
  });

  it('has z-index 1000', function() {
    expect($('.livefield-results').css('z-index').toString()).toEqual('1000');
  });

  it('is appended to body', function() {
    expect($('.livefield-results').parent()[0]).toEqual($('body')[0]);
  });

  it('matches the input width', function() {
    runs(function() {
      var inputWidth = this.$input.outerWidth();
      var resultsWidth = $('.livefield-results').outerWidth();
      expect(resultsWidth).toEqual(inputWidth);
    });
  });

  describe('onresize', function() {
    beforeEach(function() {
      this.$input.width(500);
      $(window).trigger('resize');
    });

    it('still matches width', function() {
      var inputWidth = this.$input.outerWidth();
      var resultsWidth = $('.livefield-results').outerWidth();
      expect(resultsWidth).toEqual(inputWidth);
    });
  });

  describe('if input is too close to the bottom of the page', function() {
    beforeEach(function() {
      clearsInput();
      runs(function() {
        this.$input.css({ position: 'absolute', top: 'auto', bottom: 0 })
      });
      typesCharacter();
    });

    it('positions above the input', function() {
      runs(function() {
        var inputTop = this.$input.offset().top;
        var resultsTop = $('.livefield-results').offset().top;
        var resultsBottom = resultsTop + $('.livefield-results').outerHeight() + parseInt(this.$input.css('marginTop'), 10);
        expect(resultsBottom).toEqual(inputTop);
      });
    });

    it('gives results the drop-up class', function() {
      runs(function() {
        expect('.livefield-results').toHaveClass('livefield-drop-up');
      });
    });

    describe('pressing up', function() {
      beforeEach(function() {
        pressesKey('up');
      });

      it('highlights the bottom result', function() {
        expect('.livefield-result:last-child').toBeHighlighted();
      });
    });

    describe('pressing down', function() {
      beforeEach(function() {
        pressesKey('down');
      });

      it('does nothing', function() {
        expect('.livefield-result').not.toBeHighlighted();
      });
    });
  });

});


public/livefield.spike.js
/*
 * Livefield
 */

(function($) {

// @namespace
var Livefield = {};

// @class Controller
Livefield.Controller = function(target, store) {
  var self = this;

  // constants
  var KEY_ESC   = 27,
      KEY_UP    = 38,
      KEY_DOWN  = 40,
      KEY_ENTER = 13;

  // private attrs
  var $input,
      $results,
      timeout,
      optionTemplate;

  // private methods
  var setup,
      listen,
      update,
      updateClass,
      updateOptions,
      onKeyDown,
      onFocus,
      onQueryResult,
      onStoreResponse,
      value,
      hasValue,
      moveUpOneOption,
      moveDownOneOption,
      clear,
      commit;

  // -- setup --

  setup = function() {
    $input = $(target);
    $input.addClass('livefield-input');
    $options = $('<ul class="livefield-options" />');
    $options.insertAfter($input);
    optionTemplate = Handlebars.compile($input.attr('data-option-template'));
    store = new Livefield.Store($input.attr('data-store'));
    update();
    listen();
  }

  listen = function() {
    $input.bind('keydown', onKeyDown);
    $input.bind('focus', onFocus);
  }

  // -- update --

  update = function() {
    updateClass();
    store.find(value(), onStoreResponse);
  }

  updateClass = function() {
    if (hasValue()) {
      $input.addClass('has-value');
    } else {
      $input.removeClass('has-value');
    }
  }

  updateOptions = function(options) {
    $options.html('');

    if (options.length === 0) {
      $options.hide();
    } else {
      $options.show();

      for (var i in options) {
        var option = options[i],
            $li = $(optionTemplate(option));

        $options.append($li);
      }
    }
  }

  // -- event handlers --

  onKeyDown = function(event) {
    if (event.which == KEY_ENTER) {
      event.preventDefault();
      commit();
      return;
    }

    if (event.which == KEY_UP) {
      event.preventDefault();
      moveUpOneOption();
      return;
    }

    if (event.which == KEY_DOWN) {
      event.preventDefault();
      moveDownOneOption();
      return;
    }

    if (event.which == KEY_ESC) clear();
    setTimeout(update, 0);
  }

  onStoreResponse = function(options) {
    updateOptions(options);
  }

  onFocus = function() {
    if (hasValue()) $options.show();
  }

  // -- helpers

  value = function() {
    return $input.val();
  }

  hasValue = function() {
    return value() != '';
  }

  // -- actions

  moveUpOneOption = function() {
    var $current = $options.find('.livefield-active-option');

    if ($current.length > 0) {
      var $prev = $current.prev();
      if ($prev.length > 0) {
        $current.removeClass('livefield-active-option');
        $prev.addClass('livefield-active-option');
        $input.val($prev.data('value'));
      }
    } else {
      var $prev = $options.find('.livefield-option').last();
      $prev.addClass('livefield-active-option');
      $input.val($prev.data('value'));
    }
  }

  moveDownOneOption = function() {
    var $current = $options.find('.livefield-active-option');

    if ($current.length > 0) {
      var $next = $current.next();
      if ($next.length > 0) {
        $current.removeClass('livefield-active-option');
        $next.addClass('livefield-active-option');
        $input.val($next.data('value'));
      }
    } else {
      var $next = $options.find('.livefield-option').first();
      $next.addClass('livefield-active-option');
      $input.val($next.data('value'));
    }
  }

  clear = function() {
    $input.val('');
  }

  commit = function() {
    $options.hide();
    $input.blur();
  }

  // -- run

  setup();
}

// @class Store
Livefield.Store = function(url) {
  var self = this;

  self.url = url;

  self.find = function(query, callback) {
    if (query === '') {
      callback([]);
    } else {
      callback([
        { name: "About us"               , path: "/our-company/about-us"            },
        { name: "About our suppliers"    , path: "/our-company/about-our-suppliers" },
        { name: "Roundabout maintenance" , path: "/services/roundabout-maintenance" }
      ]);
    }
  }

}

// jQuery function
$.fn.livefield = function() {
  return this.each(function() { new Livefield.Controller(this) });
}

})(jQuery);


src/livefield.js
/**
 * # Livefield.js
 *
 * http://github.com/withassociates/livefield.js
 *
 * A simple live-lookup ui library.
 *
 * ## Usage
 *
 * As a basic use case, let's say we have an input in a cms that lets users
 * specify which page a thumbnail ought to link to.
 *
 * We have an action on the server `/pages/search` that will find matching
 * pages and return them as JSON.
 *
 * This is how we'd implement livefield:
 *
 *     <!-- input to bind to -->
 *     <input
 *       type="text"
 *       name="link_path"
 *       id="my-input"
 *       data-store="/pages/search/{{query}}"
 *       data-template="#link-template"
 *     />
 *
 *     <!-- template for result options -->
 *     <script type="template/handlebars" id="link-template">
 *       <li data-value="{{path}}">
 *         <span class="name">{{name}}</span>
 *         <span class="path">{{path}}</span>
 *       </li>
 *     </script>
 *
 *     <!-- and activate livefield -->
 *     <script>
 *       $('#my-input').livefield();
 *     </script>
 *
 * ### Data-store
 *
 * Your input needs `data-store` to specify the lookup URL. This can be
 * plain text or a handlebars template. If it's a template, it will
 * receive just the `query` paramters.
 *
 * ### Data-template
 *
 * Your input also needs `data-template`. This should be a jQuery-compatible
 * selector for the element containing the template. Typically, this will
 * be in a `script` tag as in the example above.
 *
 * ### Data-value
 *
 * Your result option template needs to have the `data-value` option.
 * The value of this attribute will be used to populate your input.
 *
 * ## Dependencies
 *
 * * [jQuery](http://jquery.com/) ~> 1.5
 * * [Handlebars](http://handlebars.strobeapp.com/) ~> 1.0
 *
 * ## Contributors
 *
 * * jamie@withassociates.com
 *
 * ## License
 *
 * (The MIT License)
 *
 * Copyright (c) 2011 With Associates
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * 'Software'), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 *
 */

(function($, window) {

// module Livefield
var Livefield = { VERSION: '0.10.0' };

// class Livefield.Controller
Livefield.Controller = function(options) {
  var self = this;

  // constants
  var KEY_ESC   = 27,
      KEY_UP    = 38,
      KEY_DOWN  = 40,
      KEY_ENTER = 13;
      DEFAULT_TEMPLATE = '<li class="livefield-result" data-value="{{value}}">{{name}}</li>';

  // views
  var $input,
      $results,
      template,
      savedVal;

  // -- SETUP --

  function setup() {
    $input = $(options.input);
    self.store = new Livefield.Store({
      url: options.store || $input.attr('data-store')
    });
    setupInput();
    setupStore();
    setupTemplate();
    setupBindings();
  }

  function setupInput() {
    $input.addClass('livefield-input');
  }

  function setupStore() {
    self.store = new Livefield.Store({
      url: options.store || $input.attr('data-store')
    });
  }

  function setupTemplate() {
    var $template = $($input.attr('data-template') || DEFAULT_TEMPLATE);
    template = Handlebars.compile($template.html());
  }

  function setupBindings() {
    $input.bind('keydown keypress', onKeyDown);
    $input.bind('focus', onFocus);
    $input.bind('blur', onBlur);
  }

  // -- ACTIONS --

  function update() {
    if (hasValue()) {
      find();
    } else {
      updateResults([]);
    }
  }

  function find() {
    self.store.find(value(), onFindComplete);
  }

  function updateResults(results) {
    if (results.length === 0) {
      removeResults();
    } else {
      savedVal = value();
      appendResults();
      $results.html('');
      for (var i in results) {
        var result = results[i];
        var $result = $(template(result));
        $result.addClass('livefield-result');
        $result.appendTo($results);
      }
      positionResults();
    }
  }

  function updateValue() {
    if (!$results) return;

    var $chosen = $results.find('.livefield-highlighted').first();
    if ($chosen.length > 0) {
      if (savedVal == null) savedVal = $input.val();
      $input.val($chosen.attr('data-value'));
    } else {
      if (savedVal) {
        $input.val(savedVal);
        savedVal = null;
      }
    }
  }

  function commit() {
    $input.blur();
    removeResults();
  }

  // -- HELPERS --

  function value() {
    return $input.val();
  }

  function hasValue() {
    return value() != '';
  }

  function appendResults() {
    if (!$results) {
      $results = $('<ul class="livefield-results" />').
        delegate('.livefield-result', 'mouseout', onMouseOut).
        delegate('.livefield-result', 'mouseover', onMouseOver).
        delegate('.livefield-result', 'mousedown', onMouseDown).
        appendTo('body');
      positionResults();
      $(window).bind('resize', positionResults);
    }
  }

  function positionResults() {
    var padding       = $results.outerWidth() - $results.innerWidth(),
        resultsHeight = $results.outerHeight();

    $results.hide();

    var bodyHeight    = $('body').height(),
        inputTop      = $input.offset().top,
        inputHeight   = $input.outerHeight(),
        inputBottom   = inputTop + inputHeight,
        hasSpaceBelow = (bodyHeight - inputTop > resultsHeight),
        hasSpaceAbove = (inputTop > resultsHeight);

    if (hasSpaceBelow || !hasSpaceAbove) {
      var top = inputBottom;
      $results.removeClass('livefield-drop-up');
    } else {
      var top = inputTop - resultsHeight;
      $results.addClass('livefield-drop-up');
    }

    $results.css({
      position : 'absolute',
      zIndex   : '1000',
      width    : ($input.outerWidth() - padding) + 'px',
      left     : $input.offset().left + 'px',
      top      : top + 'px'
    }).show();
  }

  function removeResults() {
    if ($results) {
      $results.remove();
      $results = null;
      $(window).unbind('resize', positionResults);
    }
  }

  function highlightResult(which) {
    var dropUpMode = $results && $results.hasClass('livefield-drop-up');

    switch (which) {
    case 'none':
      $results.find('.livefield-highlighted').
        removeClass('livefield-highlighted');
      break;
    case 'next':
      var $current = $results.find('.livefield-highlighted');
      var $next;

      if ($current.length > 0) {
        $next = $current.next();
      } else if (!dropUpMode) {
        $next = $results.find('.livefield-result:first-child');
      }

      if ($next && $next.length > 0) {
        $current.removeClass('livefield-highlighted');
        $next.addClass('livefield-highlighted');
      } else if (dropUpMode) {
        $current.removeClass('livefield-highlighted');
      }
      break;
    case 'prev':
      var $current = $results.find('.livefield-highlighted');
      var $prev;

      if ($current.length > 0) {
        $prev = $current.prev();
      } else if (dropUpMode) {
        $prev = $results.find('.livefield-result:last-child');
      }

      if ($prev && $prev.length > 0) {
        $current.removeClass('livefield-highlighted');
        $prev.addClass('livefield-highlighted');
      } else if (!dropUpMode) {
        $current.removeClass('livefield-highlighted');
      }
      break;
    default:
      var $current = $results.find('.livefield-highlighted');
      var $chosen = $(which);
      $current.removeClass('livefield-highlighted');
      $chosen.addClass('livefield-highlighted');
    }
  }

  // -- EVENT HANDLERS --

  function onKeyDown(event) {
    if (event.which === KEY_DOWN && $results) {
      event.preventDefault();
      highlightResult('next');
      updateValue();
      return;
    }

    if (event.which === KEY_UP && $results) {
      event.preventDefault();
      highlightResult('prev');
      updateValue();
      return;
    }

    if (event.which === KEY_ENTER) {
      event.preventDefault();
      commit();
      return;
    }

    if (event.which === KEY_ESC) {
      $input.val('');
    }

    self.store.stop();
    setTimeout(update, 0); // deferred
  }

  function onMouseOver(event) {
    highlightResult(this);
    updateValue();
  }

  function onMouseOut(event) {
    highlightResult('none');
    updateValue();
  }

  function onMouseDown(event) {
    highlightResult(this);
    updateValue();
    commit();
  }

  function onFocus(event) {
    updateValue();
    update();
    setTimeout(function() {
      $input.select();
    }, 0);
  }

  function onBlur(event) {
    removeResults();
  }

  function onFindComplete(results) {
    updateResults(results);
  }

  // -- RUN --

  setup();
}

// class Livefield.Store
Livefield.Store = function(options) {
  var self = this;

  var MAX_QUEUE_LENGTH = 1;

  var transport,
      urlTemplate,
      cache = {};

  self.find = function(query, callback) {
    self.stop();
    self.queue.push([query, callback]);
    while (self.queue.length > MAX_QUEUE_LENGTH) self.queue.shift();
    if (!self.busy) nextInQueue();
  }

  self.stop = function() {
    if (transport) {
      transport.abort();
      transport = null;
    }
    self.busy = false;
  }

  function nextInQueue() {
    if (self.queue.length === 0) return;

    self.busy = true;

    var args     = self.queue.shift(),
        query    = args[0].replace(/\/+/g, ' '),
        callback = args[1];
        url      = urlTemplate.replace(':query', query);

    if (cache[url]) {
      onSuccess(url, callback);
    } else {
      transport = $.getJSON(url).then(function(data) {
        if (!self.busy) return;
        transport = null;
        cache[url] = data;
        onSuccess(url, callback);
      }, onFail);
    }
  }

  function onSuccess(url, callback) {
    callback(cache[url]);
    self.busy = false;
    nextInQueue();
  }

  function onFail() { // TODO
    // $.error('Store failed to fetch data');
    // transport = null;
    // if (self.busy) {
    //   self.busy = false;
    //   nextInQueue();
    // }
  }

  function setup() {
    urlTemplate = options.url;
    self.queue = [];
  }

  setup();
}

// jQuery plugin
$.fn.livefield = function(options) {
  return this.each(function() {
    new Livefield.Controller($.extend(options, { input: this }));
  });
}

// Make module globally available
window.Livefield = Livefield;

})(jQuery, window);
